src/tir/schedule/primitive/hide_buffer_access.cc

```diff
@@ -0,0 +1,171 @@
+           buf_type_ + " instead.";
+  }
+
+  IRModule mod() const final { return mod_; }
+  Array<ObjectRef> LocationsOfInterest() const final { return {}; }
+
+ private:
+  IRModule mod_;
+  String buf_type_;
+};
+
+class InvalidIndexError : public ScheduleError {
+ public:
+  explicit InvalidIndexError(IRModule mod, int num_access_regions, int buf_idx)
+      : mod_(std::move(mod)), num_access_regions_(num_access_regions), buf_idx_(buf_idx) {}
+
+  String FastErrorString() const final {
+    return "ScheduleError: Invalid buffer index array for hide_buffer_access schedule.";
+  }
+
+  String DetailRenderTemplate() const final {
+    return "The buffer index array for hide_buffer_access schedule should be a list of integers"
+           " between 0 and " +
+           std::to_string(num_access_regions_ - 1) + ", got " + std::to_string(buf_idx_) +
+           " instead.";
+  }
+
+  IRModule mod() const final { return mod_; }
+
+  Array<ObjectRef> LocationsOfInterest() const final { return {}; }
+
+ private:
+  IRModule mod_;
+  int num_access_regions_;
+  int buf_idx_;
+};
+
+}  // namespace
+
+/******** Implementation ********/
+
+void HideBufferAccess(ScheduleState self, const StmtSRef& block_sref, const String& buf_type,
+                      const Array<IntImm>& buf_index_array) {
+  /*!
+   * Check:
+   *   - validity of buf_index_array
+   *   - validity of buf_type
+   */
+  const BlockNode* block = TVM_SREF_TO_BLOCK(block_sref);
+  int num_access_regions = 0;
+  if (buf_type == "read") {
+    num_access_regions = block->reads.size();
+  } else if (buf_type == "write") {
+    num_access_regions = block->writes.size();
+  } else {
+    throw BufTypeError(self->mod, buf_type);
+  }
+
+  std::set<int> buf_indices;
+  for (const IntImm& buf_idx : buf_index_array) {
+    int buf_idx_val = buf_idx->value;
+    if (buf_idx_val >= 0 && buf_idx_val < num_access_regions) {
+      buf_indices.insert(buf_idx_val);
+    } else {
+      throw InvalidIndexError(self->mod, num_access_regions, buf_idx_val);
+    }
+  }
+
+  /* Step 0: Collect new buffer access regions. */
+
+  Array<BufferRegion> reads, writes;
+
+  if (buf_type == "read") {
+    for (size_t i = 0; i < block->reads.size(); ++i) {
+      if (!buf_indices.count(i)) {
+        reads.push_back(block->reads[i]);
+      }
+    }
+    writes = block->writes;
+  } else if (buf_type == "write") {
+    for (size_t i = 0; i < block->writes.size(); ++i) {
+      if (!buf_indices.count(i)) {
+        writes.push_back(block->writes[i]);
+      }
+    }
+    reads = block->reads;
+  } else {
+    CHECK(false) << "Unrecognized buffer type " << buf_type << ", only support read/write";
+  }
+
+  /* Step 1: Replace old block with the new block */
+
+  auto n = make_object<BlockNode>(*block);
+  n->reads = reads;
+  n->writes = writes;
+  Block new_block = Block(n);
+  Map<Block, Block> blk_map;
+  blk_map.Set(GetRef<Block>(block), new_block);
+  self->Replace(block_sref, new_block, blk_map);
```

Comment:
Considering that we use the `Replace` function to update the `reads/writes` for the new block, the dependence analysis would also be updated now, but the new updated dependences might not be correct as there could be missing dependences which in turn could cause correctness issues if I understand correctly.

Could we rename the primitive to `unsafe_hide_buffer_access` similar to the existing `unsafe_set_dtype` to notify users of the potential to get incorrect output?

Reply:
Yes you are right, this schedule primitive should be used as last steps (right before tensorize) during scheduling.
It makes sense to me to mark it as `unsafe`.